dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs

```diff
@@ -0,0 +1,516 @@
+        }
+        catch (HttpRequestException e)
+        {
+            this._log.LogError(e, "Vector delete request failed: {0}", e.Message);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task RemoveBatchAsync(string collectionName, IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    {
+        await Task.WhenAll(keys.Select(async k => await this.RemoveAsync(collectionName, k, cancellationToken).ConfigureAwait(false))).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(
+        string collectionName,
+        Embedding<float> embedding,
+        int limit,
+        double minRelevanceScore = 0,
+        bool withEmbeddings = false,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        this._log.LogTrace("Searching top {0} nearest vectors", limit);
+        Verify.NotNull(embedding, "The given vector is NULL");
+        string className = ToWeaviateFriendlyClassName(collectionName);
+
+        using HttpRequestMessage request = new GraphGetRequest
+        {
+            Class = className,
+            Vector = embedding.Vector,
+            Distance = minRelevanceScore,
+            Limit = limit,
+            WithVector = withEmbeddings
+        }.Build();
+
+        (HttpResponseMessage response, string responseContent) = await this.ExecuteHttpRequestAsync(request, cancellationToken).ConfigureAwait(false);
+        response.EnsureSuccessStatusCode();
+
+        GraphResponse? data = JsonSerializer.Deserialize<GraphResponse>(responseContent, s_jsonSerializerOptions);
+
+        if (data == null)
+        {
+            this._log.LogWarning("Unable to deserialize Search response");
+            yield break;
+        }
+
+        List<(MemoryRecord, double)> result = new();
+        JsonArray jsonArray = data.Data["Get"]![className]!.AsArray();
+
+        // ReSharper disable once LoopCanBeConvertedToQuery
+        foreach (JsonNode? json in jsonArray)
+        {
+            string id = json!["_additional"]!["id"]!.GetValue<string>();
+            Embedding<float> vector = Embedding<float>.Empty;
+            if (json["_additional"]!["vector"] != null)
+            {
+                IEnumerable<float> floats = json["_additional"]!["vector"]!.AsArray().Select(a => a!.GetValue<float>());
+                vector = new(floats);
+            }
+
+            string text = json["sk_text"]!.GetValue<string>();
+            string description = json["sk_description"]!.GetValue<string>();
+            string additionalMetadata = json["sk_additional_metadata"]!.GetValue<string>();
+            string key = json["sk_id"]!.GetValue<string>();
+            DateTime? timestamp = json["sk_timestamp"] != null
+                ? Convert.ToDateTime(json["sk_timestamp"]!.GetValue<string>(), CultureInfo.InvariantCulture)
+                : null;
```

Comment:
Can we describe this model as C# class and use `JsonSerializer.Deserialize`? In case it's not possible and we need to do deserialization manually, I would move this logic to separate method/class, as it's quite complex, wrap it with try/catch, in case of deserialization exception - re-throw `WeaviateMemoryException` with reason why operation failed.

Reply:
I didn't model this is as class as I felt it implied that this structure is somewhat fixed, whereas it has been created through the use of Weaviate object properties: https://github.com/microsoft/semantic-kernel/pull/1219/files#diff-6f8a37e5041a073949f0faa6e55e8ec3e6b29c36a0e8e38e0efd38bf13d1f020R42-R49

I will wrap in a try/catch and throw with a  `WeaviateMemoryException `